include/dice/template-library/integral_template_tuple.hpp

```diff
@@ -4,108 +4,307 @@
-				if constexpr (DIRECTION == Direction::up)
-					return std::make_tuple(Entry<integral_type(MIN + LENGTH - 1 - IDS)>{args...}...);
-				else
-					return std::make_tuple(Entry<integral_type(MAX - LENGTH + 1 + IDS)>{args...}...);
+	struct individual_construct_t {};
+	inline constexpr individual_construct_t individual_construct;
+
+	namespace itt_detail {
+		/**
+		 * generates the std::integer_sequence<Int, first, ..., last>
+		 * @note first is allowed to be greater than last
+		 *
+		 * @tparam Int the integral type of the std::integer_sequence
+		 * @tparam first the starting integer
+		 * @tparam last the end integer
+		 */
+		template<std::integral Int, Int first, Int last, Int ...ixs>
+		constexpr auto make_integer_sequence(std::integer_sequence<Int, ixs...> = {}) {
+			std::integer_sequence<Int, ixs..., first> const acc;
+
+			if constexpr (first == last) {
+				return acc;
+			} else if constexpr (first < last) {
+				return make_integer_sequence<Int, first + 1, last>(acc);
+			} else {
+				return make_integer_sequence<Int, first - 1, last>(acc);
+			}
+		}
+
+		/**
+		 * Selects the nth type from the given parameter pack
+		 */
+		template<size_t index, typename ...Ts>
+		struct nth_type;
+
+		template<typename First, typename ...Rest>
+		struct nth_type<0, First, Rest...> {
+			using type = First;
+		};
+
+		template<size_t index, typename First, typename ...Rest>
+		struct nth_type<index, First, Rest...> {
+			using type = typename nth_type<index - 1, Rest...>::type;
+		};
+
+		template<size_t, typename T>
+		struct struct_tuple_leaf {
+			T value_;
+
+			constexpr struct_tuple_leaf() noexcept(std::is_nothrow_default_constructible_v<T>)
+				: value_{} {
 			}
 
-			/** Wrapper for gen_tuple.
-			 * @return The constructed tuple.
-			 */
-			static auto make_tuple(Args &&...args) {
-				return gen_tuple(std::make_integer_sequence<integral_type, LENGTH>(), std::forward<Args>(args)...);
+			template<typename Arg>
+			explicit constexpr struct_tuple_leaf(individual_construct_t, Arg &&arg) noexcept(std::is_nothrow_constructible_v<T, decltype(std::forward<Arg>(arg))>)
+				: value_{std::forward<Arg>(arg)} {
 			}
 
-			/* CAUTION: has to be __after__ the make_tuple function.
-			 * Also make_tuple isn't allowed to have overloads.
-			 */
-			using type = std::invoke_result_t<decltype(make_tuple), Args &&...>;
+			constexpr auto operator<=>(struct_tuple_leaf const &other) const noexcept requires requires (T const &value) { value <=> value; } {
+				return value_ <=> other.value_;
+			}
 		};
 
-		typename tuple_generator::type count_tuple_;
+		template<typename Markers, typename ...Ts>
+		struct struct_tuple_base;
 
-	public:
-		explicit integral_template_tuple(Args &&...args)
-			: count_tuple_(tuple_generator::make_tuple(std::forward<Args>(args)...)) {}
+		template<size_t ...ixs, typename ...Ts>
+		struct struct_tuple_base<std::index_sequence<ixs...>, Ts...> : struct_tuple_leaf<ixs, Ts>... {
```

Comment:
Thanks, but I just now noticed that this is total overkill here because the types can never repeat. The only reason this exists is for somthing like `tuple<int, int>` where we somehow need to disambiguate the bases `leaf<int>` bases. Ups 😅.

Well I guess if we need it at some point, it will be here.